File: sources/net.sf.j2s.ajax/ajaxswt/net/sf/j2s/ajax/AWindowDelegate.java
/*******************************************************************************
 * Copyright (c) 2007 java2script.org and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Zhou Renjian - initial API and implementation
 *******************************************************************************/

package net.sf.j2s.ajax;

import org.eclipse.jface.window.Window;
import org.eclipse.swt.events.DisposeEvent;
import org.eclipse.swt.events.DisposeListener;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

/**
 * @author zhou renjian
 *
 * 2006-4-29
 */
public class AWindowDelegate {
	Window win;

	AWindowDelegate(Window win) {
		super();
		this.win = win;
	}
	
	public static void open(Window win,AWindowRunnable runnable) {
		win.setBlockOnOpen(false);
		new AWindowDelegate(win).nonBlockOpen(null, runnable);
	}
	/* Only JavaScript call this method. */
	static void asyncOpen(Window win, Object oThis, AWindowRunnable runnable) {
		win.setBlockOnOpen(false);
		new AWindowDelegate(win).nonBlockOpen(oThis, runnable);
	}
	
	void nonBlockOpen(final Object oThis, final AWindowRunnable runnable) {
		if (runnable instanceof AWindowRunnable) {
			runnable.setWindow(win);
		}
		boolean isJ2SEnv = false;
		/**
		 * @j2sNative
		 * isJ2SEnv = true;
		 */ {}
		if (isJ2SEnv) {
			//win.open();
			/**
			 * @j2sNative
			 * this.win.open ();
			 */ {}
			(new Runnable() {
				public void run() {
					Shell shell = win.getShell();
					if (shell == null) {
						getActiveDisplay().timerExec(10, this);
						return;
					}
					shell.addDisposeListener(new DisposeListener() {
						public void widgetDisposed(DisposeEvent e) {
							e.display.update();
							if (oThis == null) {
								e.display.timerExec(5, runnable);
							} else
							/**
							 * @j2sNative
							 * var $runnable = this.f$.runnable;
							 * var $oThis = this.f$.oThis;
							 * window.setTimeout (function () {
							 * $runnable.apply ($oThis);
							 * }, 0);
							 */
							{
							
							}
						}
					});
					shell.getDisplay().readAndDispatch();
				}
			}).run();
		} else 
			/**
			 * @j2sNative
			 */
		{
			(new Thread() {
				public void run() {
					while (win.getShell() == null) {
						try {
							Thread.sleep(10);
						} catch (InterruptedException e) {
							e.printStackTrace();
						}
					}
					getActiveDisplay().syncExec(new Runnable() {
						public void run() {
							win.getShell().addDisposeListener(new DisposeListener() {
								public synchronized void widgetDisposed(DisposeEvent e) {
									e.display.update();
									runnable.run();
								}
							});
						}
					});
				}
			}).start();
			win.open();
		}
	}
	
	Display getActiveDisplay() {
		Display display = null;
		Shell shell = win.getShell();
		if (shell != null) {
			display = shell.getDisplay();
		}
		if (display == null) {
			display = Display.getCurrent();
		}
		if (display == null) {
			display = Display.getDefault();
		}
		return display;
	}
}

File: trunk/incubator/net.sf.j2s.test.mauve/j2slib/net/sf/j2s/ajax/AWindowDelegate.js
$_J("net.sf.j2s.ajax");
$_L(null,"net.sf.j2s.ajax.AWindowDelegate",["$wt.events.DisposeListener","$wt.widgets.Display"],function(){
c$=$_C(function(){
this.win=null;
$_Z(this,arguments);
},net.sf.j2s.ajax,"AWindowDelegate");
$_K(c$,
function(win){
this.win=win;
},"org.eclipse.jface.window.Window");
c$.open=$_M(c$,"open",
function(win,runnable){
win.setBlockOnOpen(false);
new net.sf.j2s.ajax.AWindowDelegate(win).nonBlockOpen(null,runnable);
},"org.eclipse.jface.window.Window,net.sf.j2s.ajax.AWindowRunnable");
c$.asyncOpen=$_M(c$,"asyncOpen",
function(win,oThis,runnable){
win.setBlockOnOpen(false);
new net.sf.j2s.ajax.AWindowDelegate(win).nonBlockOpen(oThis,runnable);
},"org.eclipse.jface.window.Window,~O,net.sf.j2s.ajax.AWindowRunnable");
$_M(c$,"nonBlockOpen",
function(oThis,runnable){
if($_O(runnable,net.sf.j2s.ajax.AWindowRunnable)){
runnable.setWindow(this.win);
}var isJ2SEnv=false;
{
isJ2SEnv=true;
}if(isJ2SEnv){
{
this.win.open();
}((($_D("net.sf.j2s.ajax.AWindowDelegate$1")?0:net.sf.j2s.ajax.AWindowDelegate.$AWindowDelegate$1$()),$_N(net.sf.j2s.ajax.AWindowDelegate$1,this,$_F("oThis",oThis,"runnable",runnable)))).run();
}else{
}},"~O,net.sf.j2s.ajax.AWindowRunnable");
$_M(c$,"getActiveDisplay",
function(){
var display=null;
var shell=this.win.getShell();
if(shell!=null){
display=shell.getDisplay();
}if(display==null){
display=$wt.widgets.Display.getCurrent();
}if(display==null){
display=$wt.widgets.Display.getDefault();
}return display;
});
c$.$AWindowDelegate$1$=function(){
$_H();
c$=$_W(net.sf.j2s.ajax,"AWindowDelegate$1",null,Runnable);
$_M(c$,"run",
function(){
var shell=this.b$["net.sf.j2s.ajax.AWindowDelegate"].win.getShell();
if(shell==null){
this.b$["net.sf.j2s.ajax.AWindowDelegate"].getActiveDisplay().timerExec(10,this);
return;
}shell.addDisposeListener((($_D("net.sf.j2s.ajax.AWindowDelegate$1$1")?0:net.sf.j2s.ajax.AWindowDelegate.$AWindowDelegate$1$1$()),$_N(net.sf.j2s.ajax.AWindowDelegate$1$1,this,$_F("oThis",this.f$.oThis,"runnable",this.f$.runnable))));
shell.getDisplay().readAndDispatch();
});
c$=$_P();
};
c$.$AWindowDelegate$1$1$=function(){
$_H();
c$=$_W(net.sf.j2s.ajax,"AWindowDelegate$1$1",null,$wt.events.DisposeListener);
$_V(c$,"widgetDisposed",
function(e){
e.display.update();
if(this.f$.oThis==null){
e.display.timerExec(5,this.f$.runnable);
}else{
var $runnable=this.f$.runnable;
var $oThis=this.f$.oThis;
window.setTimeout(function(){
$runnable.apply($oThis);
},0);
}},"$wt.events.DisposeEvent");
c$=$_P();
};
});
